File: src/lira/linguistics/role/token_resolver.py
"""TokenResolver: the read path's own tokenizer + candidate-resolution
step (Linguistics Layer developer specification, 8.8). Wraps
LinguisticLexer (sentence/token splitting, unchanged) and
GraphProcessor.process_token_candidates (candidate resolution, unchanged
identify_word integration) into the TokenReading sequences
Phrase.read()/Clause.read()/Sentence.read() consume -- this is the "text
or pre-resolved tokens" split spec 14.3 asks Sentence.read() to accept:
raw text always passes through here first, and always through the same
LinguisticLexer the write path uses, so a token or sentence boundary
never differs between LinguisticController.tokenize_prompt (write) and
LinguisticController.read_text (read) for the same input."""

from typing import Tuple

from ..data.token_reading import TokenReading
from .grammar_configurator import GrammarConfigurator
from .graph_processor import GraphProcessor
from .lexer import LinguisticLexer


class TokenResolver:
    def __init__(self, graph_processor: GraphProcessor):
        self.graph_processor = graph_processor

    def resolve_sentence(
        self, raw_sentence_text: str, *, sentence_index: int = 0,
    ) -> Tuple[TokenReading, ...]:
        """One sentence's worth of TokenReadings, in order -- every
        seeded candidate retained per token (spec 7's "candidate parts
        of speech"), none collapsed to a single sense the way
        process_token's materialisation step does."""
        raw_tokens = LinguisticLexer.extract_tokens(raw_sentence_text)
        return tuple(
            self.graph_processor.process_token_candidates(
                token_text,
                sentence_index=sentence_index, token_index=idx, is_sentence_start=(idx == 0),
                preceding_words=tuple(raw_tokens[:idx]), following_words=tuple(raw_tokens[idx + 1:]),
            )
            for idx, token_text in enumerate(raw_tokens)
        )

    def resolve_text(
        self, raw_text: str, *, grammar: GrammarConfigurator,
    ) -> Tuple[Tuple[TokenReading, ...], ...]:
        """One tuple of TokenReadings per sentence in `raw_text`, split
        by the same LinguisticLexer.split_sentences the write path uses.
        Sentence.read(text=...) uses this when handed a raw string;
        given an already-resolved TokenReading sequence instead, it
        skips this method entirely (spec 14.3's "text or pre-resolved
        tokens")."""
        raw_sentences = LinguisticLexer.split_sentences(raw_text, grammar)
        return tuple(
            self.resolve_sentence(sentence_text, sentence_index=idx)
            for idx, sentence_text in enumerate(raw_sentences)
            if sentence_text
        )

File: prototype/src/lira/linguistics/role/token_resolver.ts
import type { TokenReading } from "../data/token_reading";
import type { GrammarConfigurator } from "./grammar_configurator";
import type { GraphProcessor } from "./graph_processor";
import { LinguisticLexer } from "./lexer";

/** TokenResolver: the read path's own tokenizer + candidate-resolution
 * step (Linguistics Layer developer specification, 8.8). Wraps
 * LinguisticLexer (sentence/token splitting, unchanged) and
 * GraphProcessor.processTokenCandidates (candidate resolution,
 * unchanged identifyWord integration) into the TokenReading sequences
 * readPhrase()/readClause()/readSentence() consume -- this is the "text
 * or pre-resolved tokens" split spec 14.3 asks Sentence.read() to
 * accept: raw text always passes through here first, and always
 * through the same LinguisticLexer the write path uses, so a token or
 * sentence boundary never differs between LinguisticController.tokenizePrompt
 * (write) and LinguisticController.readText (read) for the same input.
 *
 * Ported from linguistics/role/token_resolver.py. */
export class TokenResolver {
  constructor(private readonly graphProcessor: GraphProcessor) {}

  /** One sentence's worth of TokenReadings, in order -- every seeded
   * candidate retained per token (spec 7's "candidate parts of
   * speech"), none collapsed to a single sense the way processToken's
   * materialisation step does. Uses GraphProcessor.processPhraseCandidates,
   * not processTokenCandidates directly: a closed-class multi-word entry
   * like "in spite of" resolves as one TokenReading spanning three raw
   * tokens (its own tokenSpan) rather than three independent
   * single-word readings, so the cursor below advances by tokenSpan,
   * not by 1. */
  resolveSentence(rawSentenceText: string, sentenceIndex = 0): readonly TokenReading[] {
    const rawTokens = LinguisticLexer.extractTokens(rawSentenceText);
    const readings: TokenReading[] = [];
    let rawIndex = 0;
    while (rawIndex < rawTokens.length) {
      const reading = this.graphProcessor.processPhraseCandidates(rawTokens, rawIndex, {
        sentenceIndex, isSentenceStart: rawIndex === 0,
      });
      readings.push(reading);
      rawIndex += reading.tokenSpan;
    }
    return readings;
  }

  /** One array of TokenReadings per sentence in `rawText`, split by the
   * same LinguisticLexer.splitSentences the write path uses.
   * readSentence(text=...) uses this when handed a raw string; given an
   * already-resolved TokenReading sequence instead, it skips this
   * method entirely (spec 14.3's "text or pre-resolved tokens"). */
  resolveText(rawText: string, grammar: GrammarConfigurator): readonly (readonly TokenReading[])[] {
    const rawSentences = LinguisticLexer.splitSentences(rawText, grammar);
    return rawSentences
      .filter((sentenceText) => sentenceText)
      .map((sentenceText, idx) => this.resolveSentence(sentenceText, idx));
  }
}
